### ISSUE TITLE
[MRESOLVER-7] Download dependency POMs in parallel in BF collector

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-resolver#817]
**[Harald Wellmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hwellmann)** opened **[MRESOLVER-7](https://issues.apache.org/jira/browse/MRESOLVER-7?redirect=false)** and commented

### Background

When building a project with dependencies not yet available in the local repository, I noticed that Maven 3.3.9/Aether 1.0.2 first downloads the dependency POMs _sequentially_ and then proceeds downloading the dependency JARs with up to 5 threads _in parallel_.

Due to this, when first building a project with a large number of dependencies, downloading a large number of small POMs may take a lot longer than downloading the much larger JARs, or even longer than building the project itself, especially when a repository manager is used which increases the download latency.

### Enhancement

Download POMs of (transitive) dependencies in parallel to significantly speed up initial builds of large projects.

---

**Affects:** Aether 1.0.2

**Attachments:**
- [resolve_deps.png](https://issues.apache.org/jira/secure/attachment/13047919/resolve_deps.png) (_38.74 kB_)
- [resolver.log](https://issues.apache.org/jira/secure/attachment/13038040/resolver.log) (_14.49 kB_)

**Issue Links:**
- [MNG-5896](https://issues.apache.org/jira/browse/MNG-5896) Download dependency POMs in parallel
 (_**"blocks"**_)
- [MNG-6958](https://issues.apache.org/jira/browse/MNG-6958) mvn only uses a single thread for downloading poms.
 (_**"is duplicated by"**_)
- [MNG-6672](https://issues.apache.org/jira/browse/MNG-6672) Upgrade Maven Resolver to 1.4.2
 (_**"is part of"**_)
- [MRESOLVER-92](https://issues.apache.org/jira/browse/MRESOLVER-92) Revert MRESOLVER-7
 (_**"causes"**_)
- [MNG-6712](https://issues.apache.org/jira/browse/MNG-6712) Downgrade maven-resolver:1.4.0 to 1.3.3

- [MRESOLVER-32](https://issues.apache.org/jira/browse/MRESOLVER-32) Support parallel artifact/metadata uploads

**Remote Links:**
- [GitHub Pull Request #10
](https://github.com/apache/maven-resolver/pull/10)
- [GitHub Pull Request #142
](https://github.com/apache/maven-resolver/pull/142)
- [GitHub Pull Request #178
](https://github.com/apache/maven-resolver/pull/178)
- [GitHub Pull Request #2
](https://github.com/apache/maven-resolver/pull/2)
- [GitHub Pull Request #30
](https://github.com/apache/maven-resolver/pull/30)

24 votes, 34 watchers

**Comment by jira-importer** [comment]
**[Harald Wellmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hwellmann)** commented

Pull request:

https://github.com/apache/maven-resolver/pull/2

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Please raise the review on the dev mailing list, we'd like to push a release this year.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Master is currently broken.

**Comment by jira-importer** [comment]
**[Sylwester Lachiewicz](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slachiewicz)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) any plan to review? Maybe include this one with new Maven 3.6?

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

[Sylwester Lachiewicz](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slachiewicz) Can you take a look at it cause I want to start a release within the next few days...?

**Comment by jira-importer** [comment]
**[Sylwester Lachiewicz](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slachiewicz)** commented

Please go without this fix, I'll try to work on this in next year :D

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Ok Thanks Sylwester for the feedback.

**Comment by jira-importer** [comment]
**[Harald Wellmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hwellmann)** commented

Rebased Sylwester's changes in a new PR https://github.com/apache/maven-resolver/pull/30 and fixed the failing tests.

Added a new configuration property `aether.artifact.descriptor.threads` for the number of threads used for parallel download of POMs.

The default is 5. Setting this to 1 reverts to the previous behaviour.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [9d3b324e82d9b934bf51bb8818c83caf776e6faf](https://gitbox.apache.org/repos/asf?p=mavenre-resolver.git;a=commit;h=9d3b324e82d9b934bf51bb8818c83caf776e6faf).

**Comment by jira-importer** [comment]
**[Tibor Digana](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tibor17)** commented

[Harald Wellmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hwellmann)
We have to revert this feature in resolver and resolver:1.4.0 in Maven. It should be reviewed in a separate branch.

As an explanation, our Maven builds started failing after resolver:1.4.0 has been applied.
To be unbiased and fully logical, I investigated the issue and the feature on ASF Slack. It was me who found several problems in the code, root cause of hanging Maven and deadlock of thread-resources starvation, problems with thread-unsafe classes, overrided objects across threads, one unclear part in the algorithm. And I also found that the algorithm crashed 11 unit tests in `maven-resolver` after i removed ThreadPoolExecutor (in a playgrounded non-master branch) and thus made it single-threaded.

I am sorry and not glad to say these things, but I would like you and ASF Maven to make code review of this feature, where every question and line must be clarified. Are you open for attending in such code review?

I may copy paste the text from Slack regarding our findings.
But currently we are responsible to solve our internal problems with failing builds.
The Jenkins builds have been failing over one month!

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

This change has been reverted and moved to the branch `pre-MRESOLVER-92`.

**Comment by jira-importer** [comment]
**[Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) Does MRESOLVER-123 fix some of the issues which appeared with https://github.com/apache/maven-resolver/pull/30?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin), to be honest, no idea. Maybe, you can try yourself. This is currently out of scope for me.

**Comment by jira-importer** [comment]
**[Rohit Gupta](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rajjaiswalsaumya)** commented

 When this will make way in maven ?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Rohit Gupta](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rajjaiswalsaumya), when you provide a decent PR.

**Comment by jira-importer** [comment]
**[Michael](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ffatheranderson)** commented

I regret I can not like the previous comment. :) [Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) cheers.

**Comment by jira-importer** [comment]
**[Palash Das](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ppanda-beta)** commented

You people keep fighting over decency of a PR or deadlocks of multi threaded downloads, but in reality this forces us to pay free money to the cloud providers for CI time. 
When we have 

1. sufficient network bandwidth
2.  High speed SSD
3. And good amount of cpu

 

**we still have to pay 3 hrs of CI time just to download 3.4GBs of jar files.**
Great collaboration skills guys. Great appreciations towards the OSS efforts that people show on their PRs. Approve the PR, merge it and revert it within 1 month. Awesome reward received. 

Time will come when we will switch the build tool instead of a dependency plugin. Cheers to gradle / bazel. Those comment fights will be remembered!

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Palash Das](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ppanda-beta), if you really need this feature then hire a contractor, pay him €€€ and we will review his PR. You obviously don't understand that we are volunteers don't have to do nothing for you!

**Comment by jira-importer** [comment]
**[Ivan Babiankou](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ibabiankou)** commented

I opened [a PR](https://github.com/apache/maven-resolver/pull/142) to implement this change. Unit and integration tests pass, however there seems to be a race condition (see stack trace [^resolver.log] ) that is hard to reproduce.

**Comment by jira-importer** [comment]
**[Philippe Cloutier](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=chealer)** commented

[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas) : Does this specifically affect the BF collector as opposed to other collectors? Is there a workaround?

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

[Chealer](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Chealer) the requirements to make this happen requires several changes, and one of them was done as part of breadth-first collecting of dependencies (BF) while so far "original" collector and default as well is depth-first (DF). The provided PR enables the new BF collector to download POMs in parallel, while the "original" DF is still unchanged...

But that said, I think there may be a slight mixup here: the PR was done as part of this issue (while it modified BF collector only, something that when this issue was created did not even exist yet), so to say, this issue got "hijacked"?

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

Also, have to remark that in case of collection "POM downloads" happens in two distinct cases (but interleaved during collection):
* a) when downloading the dependency POM (for example D depends on D1, D2, D3, then corresponding D1.pom, D2.pom, D3.pom will be downloaded) – these can and are downloaded in parallel with provided PR for BF collector.
* b) but, when D1.pom is being built (in model builder), and IF D1.pom has D1-parent.pom, amd parent has D1-parent-parent.pom (and so on, but similar for POM  imports, etc), these are still sequentially downloaded, as model bullder processes/builds the D1.pom. Model builder was not altered by PR.

The PR implements case a) in BF collector only. DF collector and model builder is unchanged.

**Comment by jira-importer** [comment]
**[Philippe Cloutier](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=chealer)** commented

Thank you very much for your work as well as for this valuable explanation [Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas) 

I know nothing about collectors and the model builder nuance, but if I understand the situation correctly, I recommend to:
* Restore this ticket's original title and IN PROGRESS status
* If a ticket specifically about the BF collector is needed or useful:
  * create one or clone this one
  * set it as a dependency of this one
  * mark it as resolved
* If a ticket specifically about the DF collector is needed or useful:
  * create one or clone this one
  * set it as a dependency of this one
* Ensure that the assignee of each ticket is representative

This may have been overdue, but seeing how much work was already needed, congratulations for what was already accomplished.

**Comment by jira-importer** [comment]
**[wei cai](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wecai)** commented

BF collector is a combination of 
https://github.com/apache/maven-resolver/pull/158
https://github.com/apache/maven-resolver/pull/178

It includes some performance improvements based on what we've done in our company to speed up maven build. So far the BF and original DF algorithm co-exists and the DF is still the default.

When maven 3.9.x ships with maven-resolver 1.9.0, then users can simply put a [.mvn/maven.config](https://maven.apache.org/configure.html) within project root with below content:
"-Daether.dependencyCollector.impl=bf" to enable these features:
* Skipper to skip unnecessary resolution
* Parallel descriptor files (pom, metadata.xml) downloading

Not sure what's the plan to make the BF as default one. All related Jiras about parallel pom downloading can be closed accordingly until BF becomes the default option.

**Comment by jira-importer** [comment]
**[Philippe Cloutier](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=chealer)** commented

Thank you very much [wei cai](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wecai) for these details. I would appreciate a clarification of what "Skipper to skip unnecessary resolution" means.

**Comment by jira-importer** [comment]
**[wei cai](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wecai)** commented

[Chealer](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Chealer) 

You may find more details in https://issues.apache.org/jira/browse/MRESOLVER-247 about the BF + Skipper.

**Comment by jira-importer** [comment]
**[Björn Michael](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mtomcat_daobi)** commented

Is it possible to set it in `pom.xml` as well? I'd like to put it into a parent POM.

```xml
<properties>
    <aether.dependencyCollector.impl>bf</aether.dependencyCollector.impl>
</properties>
```

**Comment by jira-importer** [comment]
**[Jonathan Haber](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jhaber)** commented

We started rolling out Maven 3.9.0 with parallel downloads to our engineering team, and the positive response has been overwhelming. I included a picture below of some of the Slack reactions the announcement received in the first few hours. Just wanted to give a big thanks to [wei cai](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wecai) and everyone else who worked on this

!https://i.imgur.com/mkGNFmY.png!

**Comment by jira-importer** [comment]
**[Philippe Cloutier](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=chealer)** commented

Thank you [wei cai](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wecai) 

Considering the change in this ticket's scope, I suggest those who voted for this ticket hoping to see a change in default behavior to also vote for MNG-5896.

**Comment by jira-importer** [comment]
**[wei cai](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wecai)** commented

[Jonathan Haber](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jhaber) 

Thanks for your feedback! Really glad it helps your Company and Team.

[Chealer](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Chealer) 

The new changes all go to BF by default as the resolver is very critical to the project build & deploy result.  And as discussed earlier before with Maven boards, the plan is to making BF as a default option in future versions based on more community feedback.

Please try the BF option and share your feedback to:

https://issues.apache.org/jira/browse/MRESOLVER-324

**Comment by jira-importer** [comment]
**[Knut Wannheden](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=knut)** commented

Yes, the new parallel collector is working very nicely!

In one build I did see an error, which I was thinking might be related to using the BF collector, but I haven't been able to reproduce it again. In my settings.xml I have multiple Maven repositories configured which "overlap" (i.e. some artifacts are contained in multiple of these repos) and I have never been able to figure out how Maven decides from which repo to download an artifact (yes, I know this is documented, but it doesn't appear to match my observations). I don't know if this extra information is relevant or helpful, but anyway, here is the error I saw. Please let me know if you want me to create an issue.

```
[ERROR] Failed to execute goal org.jacoco:jacoco-maven-plugin:0.8.8:prepare-agent (default-prepare-agent) on project glu-test-contract-service-parent: Execution default-prepare-agent of goal org.jacoco:jacoco-maven-plugin:0.8.8:prepare-agent failed: Plugin org.jacoco:jacoco-maven-plugin:0.8.8 or one of its dependencies could not be resolved: Failed to collect dependencies at org.jacoco:jacoco-maven-plugin:jar:0.8.8 -> org.apache.maven.shared:file-management:jar:1.2.1 -> org.apache.maven:maven-plugin-api:jar:2.0.6: java.io.UncheckedIOException: java.nio.file.AccessDeniedException: c:\dev\.m2-nexus\repository\org\apache\maven\maven-parent\5\maven-parent-5.pom.lastUpdated.7796060141987260961.tmp -> c:\dev\.m2-nexus\repository\org\apache\maven\maven-parent\5\maven-parent-5.pom.lastUpdated -> [Help 1]
```

**Comment by jira-importer** [comment]
**[wei cai](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wecai)** commented

[Knut Wannheden](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=knut) 

Thanks for your feedback! This is a closed JIRA, it's not appropriate to put new comments here, let's discuss about BF + Skipper + Parallel downloading in https://issues.apache.org/jira/browse/MRESOLVER-324

For your reported issue, it is not BF related, should be related with https://issues.apache.org/jira/browse/MRESOLVER-325
